core/test/gtest/environments.hpp

```diff
@@ -0,0 +1,220 @@
+            sycl_device_id = gko::experimental::mpi::map_rank_to_device_id(
+                MPI_COMM_WORLD,
+                std::max(gko::DpcppExecutor::get_num_devices("gpu"), 1));
+        }
+#endif
+
+        auto rs_count_env = std::getenv("CTEST_RESOURCE_GROUP_COUNT");
+        auto rs_count = rs_count_env ? std::stoi(rs_count_env) : 0;
+        if (rs_count == 0) {
+            std::cerr << "Running without CTest ctest_resource configuration"
+                      << std::endl;
+            return;
+        }
+        if (rs_count != size) {
+            GKO_INVALID_STATE("Invalid resource group count: " +
+                              std::to_string(rs_count));
+        }
+
+        // parse CTest ctest_resource group descriptions
+        if (rank == 0) {
+            std::cerr << "Running with CTest ctest_resource configuration:"
+                      << std::endl;
+        }
+        // OpenMP CPU threads
+        if (auto rs_omp_env = get_ctest_group("cpu", rank)) {
+            auto resource = parse_ctest_resources(rs_omp_env);
+            omp_threads = resource.slots;
+            if (rank == 0) {
+                std::cerr << omp_threads << " CPU threads" << std::endl;
+            }
+        }
+        // CUDA GPUs
+        if (auto rs_cuda_env = get_ctest_group("cudagpu", rank)) {
+            auto resource = parse_ctest_resources(rs_cuda_env);
+            cuda_device_id = resource.id;
+            std::cerr << "Rank " << rank << ": CUDA device " << cuda_device_id
+                      << std::endl;
+        }
+        // HIP GPUs
+        if (auto rs_hip_env = get_ctest_group("hipgpu", rank)) {
+            auto resource = parse_ctest_resources(rs_hip_env);
+            hip_device_id = resource.id;
+            std::cerr << "Rank " << rank << ": HIP device " << cuda_device_id
+                      << std::endl;
+        }
+        // SYCL GPUs (no other devices!)
+        if (auto rs_sycl_env = get_ctest_group("syclgpu", rank)) {
+            auto resource = parse_ctest_resources(rs_sycl_env);
+            sycl_device_id = resource.id;
+            std::cerr << "Rank " << rank << ": SYCL device " << cuda_device_id
+                      << std::endl;
+        }
+    }
+
+    static int omp_threads;
+    static int cuda_device_id;
+    static int hip_device_id;
+    static int sycl_device_id;
+};
+
+
+#ifdef GKO_COMPILING_OMP
+
+class OmpEnvironment : public ::testing::Environment {
+public:
+    void SetUp() override
+    {
+        if (ResourceEnvironment::omp_threads > 0) {
+            omp_set_num_threads(ResourceEnvironment::omp_threads);
+        }
+    }
+};
+
+#else
+
+
+class OmpEnvironment : public ::testing::Environment {};
+
+#endif
+
+
+#ifdef GKO_COMPILING_CUDA
+
+class CudaEnvironment : public ::testing::Environment {
+public:
+    void TearDown() override
+    {
+        gko::kernels::cuda::reset_device(ResourceEnvironment::cuda_device_id);
+    }
```

Comment:
Same with the device ID and maybe the device name (cudaGetDeviceProperties, might need a wrapper inside ginkgo_cuda)